src/main/kotlin/org/rust/cargo/util/CargoCommandCompletionProvider.kt

```diff
@@ -0,0 +1,257 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.util
+
+import com.intellij.codeInsight.completion.CodeCompletionHandlerBase
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.completion.CompletionType
+import com.intellij.codeInsight.completion.PrioritizedLookupElement
+import com.intellij.codeInsight.lookup.CharFilter
+import com.intellij.codeInsight.lookup.LookupElement
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.openapi.externalSystem.service.execution.cmd.ParametersListLexer
+import com.intellij.util.TextFieldCompletionProvider
+import com.intellij.util.execution.ParametersListUtil
+import org.rust.cargo.project.workspace.CargoWorkspace
+import org.rust.cargo.project.workspace.PackageOrigin
+import org.rust.lang.core.completion.addSuffix
+
+class CargoCommandCompletionProvider(
+    private val workspace: CargoWorkspace?
+) : TextFieldCompletionProvider() {
+
+    override fun getPrefix(currentTextPrefix: String): String = splitContextPrefix(currentTextPrefix).second
+
+    override fun acceptChar(c: Char): CharFilter.Result? =
+        if (c == '-') CharFilter.Result.ADD_TO_PREFIX else null
+
+    override fun addCompletionVariants(text: String, offset: Int, prefix: String, result: CompletionResultSet) {
+        val (ctx, _) = splitContextPrefix(text)
+        result.addAllElements(complete(ctx))
+    }
+
+    // public for testing
```

Comment:
Hm, generally this sounds like an excellent approach, but I am afraid I am too lazy in the end :)